plugins/jabra-file/fu-jabra-file-device.c

```diff
@@ -0,0 +1,671 @@
+		fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_ERASE, 1, "first-block");
+		fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_WRITE, 89, "next-block");
+		fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_WRITE, 10, "update-device");
+
+		/* file not on device, transfer it */
+		fw = fu_firmware_get_bytes(firmware, error);
+		if (fw == NULL)
+			return FALSE;
+		chunks = fu_chunk_array_new_from_bytes(fw, 0x00, chunk_size);
+
+		if (!fu_jabra_file_device_write_delete_file(self, error))
+			return FALSE;
+		if (!fu_jabra_file_device_write_first_block(self, firmware, error))
+			return FALSE;
+		fu_progress_step_done(progress);
+		if (!fu_jabra_file_device_write_blocks(self,
+						       chunks,
+						       fu_progress_get_child(progress),
+						       error))
+			return FALSE;
+		fu_progress_step_done(progress);
+		if (!fu_jabra_file_device_file_checksum(self, firmware_checksum, &match, error))
+			return FALSE;
+		if (!match) {
+			g_set_error(error,
+				    FWUPD_ERROR,
+				    FWUPD_ERROR_INTERNAL,
+				    "error transferring file to device, checksum doesn't match");
+			return FALSE;
+		}
+	} else {
+		/* file already on device */
+		fu_progress_set_id(progress, G_STRLOC);
+		fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_WRITE, 100, "update-device");
+	}
+
+	if (!fu_jabra_file_device_check_device_busy(self, error))
+		return FALSE;
+	if (!fu_jabra_file_device_start_update(self, error))
+		return FALSE;
+	fu_progress_step_done(progress);
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_jabra_file_device_attach(FuDevice *device, FuProgress *progress, GError **error)
+{
+	FuJabraFileDevice *self = FU_JABRA_FILE_DEVICE(device);
+	fu_device_sleep_full(FU_DEVICE(self), 900000, progress);
+	fu_device_set_remove_delay(FU_DEVICE(self), 10000);
+	fu_device_add_flag(FU_DEVICE(self), FWUPD_DEVICE_FLAG_WAIT_FOR_REPLUG);
+	return TRUE;
+}
+
+static void
+fu_jabra_file_device_set_progress(FuDevice *self, FuProgress *progress)
+{
+	fu_progress_set_id(progress, G_STRLOC);
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_RESTART, 1, "detach");
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_WRITE, 85, "write");
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_RESTART, 7, "attach");
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_BUSY, 7, "reload");
+}
+
+static void
+fu_jabra_file_device_init(FuJabraFileDevice *self)
+{
+	self->address = 0x01;
+	fu_device_add_flag(FU_DEVICE(self), FWUPD_DEVICE_FLAG_UPDATABLE);
+	fu_device_add_flag(FU_DEVICE(self), FWUPD_DEVICE_FLAG_SIGNED_PAYLOAD);
+	fu_device_add_flag(FU_DEVICE(self), FWUPD_DEVICE_FLAG_DUAL_IMAGE);
+	fu_device_add_flag(FU_DEVICE(self), FWUPD_DEVICE_FLAG_SELF_RECOVERY);
+	fu_device_add_flag(FU_DEVICE(self), FWUPD_DEVICE_FLAG_ADD_COUNTERPART_GUIDS);
+	fu_device_add_internal_flag(FU_DEVICE(self), FU_DEVICE_INTERNAL_FLAG_ONLY_WAIT_FOR_REPLUG);
+	fu_device_set_remove_delay(FU_DEVICE(self), 120000);
+	fu_device_add_protocol(FU_DEVICE(self), "com.jabra.file");
```

Comment:
Hi @hughsie , I don't recall what the process was to update LVFS to allow the cabinet files for the new protocol to be uploaded. I can see that the list (https://fwupd.org/lvfs/docs/metainfo/protocol) doesn't yet include "com.jabra.file", and I couldn't upload the firmware either since the metainfo entry for the protocol has no match.

Reply:
Oopa, I forgot to add it. In general you need to file a ticket in https://gitlab.com/fwupd/lvfs-website/-/issues  but here is fine too. Some questions: are the images signed? If signed, signed with what? e.g. SHA256+RSA-1024 -- thanks.

Reply:
It is signed, and it uses a certification based signing, with a SHA256 key